pkg/nodediscovery/nodediscovery.go

```diff
@@ -172,8 +175,44 @@ func (n *NodeDiscovery) JoinCluster(nodeName string) {
 // start configures the local node and starts node discovery. This is called on
 // agent startup to configure the local node based on the configuration options
 // passed to the agent. nodeName is the name to be used in the local agent.
-func (n *NodeDiscovery) StartDiscovery(nodeName string) {
-	n.LocalNode.Name = nodeName
+func (n *NodeDiscovery) StartDiscovery() {
+	n.localNodeLock.Lock()
+	defer n.localNodeLock.Unlock()
```

Comment:
Is this mutex needed on v1.9? `StartDiscovery` cannot be called more than once anyway and it doesn't seem to be used anywhere else.

On v1.9, we also cannot really protect writes to `LocalNode`, as it is exported.

Reply:
Oh, I missed the fact that the watcher is calling `fillLocalNode` indirectly. I think that might be inherently racy, as the fields could be written to while some other code is reading it.